File: Data Structure/Hash/Medium/LET_no347_Top K Frequent Elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        memo = collections.Counter(nums)
        n = len(nums)
        freqs = [[] for _ in range(n + 1)]
        for num, freq in memo.items():
            freqs[freq].append(num)
        
        res = []
        for f in range(n, 0, -1):
            if freqs[f]:
                res.extend(freqs[f])
                if len(res) >= k:
                    return res[:k]
        
        return res[:k]

class Solution2:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        memo = collections.Counter(nums)
        num_freq_pairs = [(num, freq) for num, freq in memo.items()]
        self.quickSelect(num_freq_pairs, 0, len(num_freq_pairs) - 1, k - 1)
        return [num for num, _ in num_freq_pairs[:k]]
    
    def quickSelect(self, pairs, left, right, target):
        if left >= right:
            return
        
        i, j = left, right
        pivot = pairs[random.randint(i, j)][1]
        while i <= j:
            while i <= j and pairs[i][1] > pivot:
                i += 1
            while i <= j and pairs[j][1] < pivot:
                j -= 1
            if i <= j:
                pairs[i], pairs[j] = pairs[j], pairs[i]
                i += 1
                j -= 1
                
        if target <= j:
            self.quickSelect(pairs, left, j, target)
        elif i <= target:
            self.quickSelect(pairs, i, right, target)

class Solution3:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        memo = collections.Counter(nums)
        h = []
        for val, freq in memo.items():
            heapq.heappush(h, (freq, val))
            if len(h) > k:
                heapq.heappop(h)
            
        return [val for freq, val in h]

File: Data Structure/Hash/Medium/LET_no347_Top K Frequent Elements.cpp
// priority queue
class Solution1 {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> memo;
        for (int num : nums)
            memo[num]++;

        priority_queue<pair<int, int>> pq;
        for (auto it = memo.begin(); it != memo.end(); it++) {
            pq.push({-it->second, it->first});
            if (pq.size() > k)
                pq.pop();
        }

        vector<int> res;
        while (!pq.empty()) {
            res.push_back(pq.top().second);
            pq.pop();
        }

        return res;
    }
};

// bucket sort
class Solution2 {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> memo;
        for (int num : nums)
            memo[num]++;

        int n = nums.size();
        vector<vector<int>> bucket(n + 1, vector<int>());
        for (auto it = memo.begin(); it != memo.end(); it++)
            bucket[it->second].push_back(it->first);

        vector<int> res;
        int i = n;
        while (k > 0) {
            while (bucket[i].empty())
                i--;
            for (int j = 0; j < bucket[i].size() && k > 0; j++, k--)
                res.push_back(bucket[i][j]);
            i--;
        }

        return res;
    }
};

